lib/logstash/inputs/DynamoDBLogParser.rb

```diff
@@ -0,0 +1,164 @@
+
+    @hash_template["dynamodb"] = Hash.new
+    @hash_template["dynamodb"]["keys"] = Hash.new
+    size_bytes = calculate_key_size_in_bytes(log)
+    @key_schema.each { |x|
+      @hash_template["dynamodb"]["keys"][x] = data_hash[x]
+    }
+    unless @view_type == "keys_only"
+      size_bytes += new_image_size
+      @hash_template["dynamodb"]["newImage"] = data_hash
+    end
+    @hash_template["dynamodb"]["sequenceNumber"] = "0"
+    @hash_template["dynamodb"]["sizeBytes"] = size_bytes
+    @hash_template["dynamodb"]["streamViewType"] = @view_type.upcase
+
+    return parse_view_type(@hash_template)
+  end
+
+  public
+  def parse_stream(log)
+    return parse_view_type(JSON.parse(@mapper.writeValueAsString(log))["internalObject"])
+  end
+
+  private
+  def calculate_key_size_in_bytes(record)
+    key_size = 0
+    @key_schema.each { |x|
+      key_size += x.length
+      value = record.get(x)
+      if not value.getB().nil?
+        b = value.getB();
+        key_size += Base64.decode64(b).length
+      elsif not value.getS().nil?
+        s = value.getS();
+        key_size += s.length;
+      elsif not value.getN().nil?
+        key_size += MAX_NUMBER_OF_BYTES_FOR_NUMBER;
+      end
+    }
+    return key_size
+  end
+
+  private
+  def parse_view_type(hash)
+    if @log_format == LogStash::Inputs::DynamoDB::LF_PLAIN
+      return hash.to_json
+    end
+    case @view_type
+    when LogStash::Inputs::DynamoDB::VT_KEYS_ONLY
+      return parse_format(hash["dynamodb"]["keys"])
+    when LogStash::Inputs::DynamoDB::VT_OLD_IMAGE
+      return parse_format(hash["dynamodb"]["oldImage"])
+    when LogStash::Inputs::DynamoDB::VT_NEW_IMAGE
+      return parse_format(hash["dynamodb"]["newImage"]) #check new and old, dynamodb.
+    end
+  end
+
+  private
+  def parse_format(hash)
+    if @log_format == LogStash::Inputs::DynamoDB::LF_DYNAMODB
+      return hash.to_json
+    else
+      return dynamodb_to_json(hash)
+    end
+  end
+
+  private
+  def dynamodb_to_json(hash)
+    return formatAttributeValueMap(hash).to_json
+  end
+
+  private
+  def formatAttributeValueMap(hash)
+    keys_to_delete = []
+    hash.each do |k, v|
+      dynamodb_key = v.keys.first
+      dynamodb_value = v.values.first
+      if @log_format == LogStash::Inputs::DynamoDB::LF_JSON_NO_BIN and (dynamodb_key == "BS" or dynamodb_key == "B")
+        keys_to_delete.push(k) # remove binary values and binary sets
+        next
+      end
+      hash[k] = formatAttributeValue(v.keys.first, v.values.first)
+    end
+    keys_to_delete.each {|key| hash.delete(key)}
+    return hash
+  end
+
+  private
+  def formatAttributeValue(key, value)
+    case key
+    when "M"
+      formatAttributeValueMap(value)
+    when "L"
+      value.map! do |v|
+        v = formatAttributeValue(v.keys.first, v.values.first)
+      end
+    when "NS","SS","BS"
+      value.map! do |v|
+        v = formatAttributeValue(key[0], v)
+      end
+      return value
```

Comment:
this return value might not be strictly necessary here as you did in the previous case.